File: go-backend-interview-prep/09_performance_and_profiling/07_strings_bytes_and_builders/main.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

func concatPlus(n int) string {
	s := ""
	for i := 0; i < n; i++ {
		s += "item" + strconv.Itoa(i) + ","
	}
	return s
}

func concatBuilder(n int) string {
	var b strings.Builder
	b.Grow(n * 10) // estimate capacity
	for i := 0; i < n; i++ {
		b.WriteString("item")
		b.WriteString(strconv.Itoa(i))
		b.WriteByte(',')
	}
	return b.String()
}

func concatJoin(n int) string {
	parts := make([]string, n)
	for i := 0; i < n; i++ {
		parts[i] = "item" + strconv.Itoa(i)
	}
	return strings.Join(parts, ",")
}

// byteStringConversion shows the cost of converting between types.
func byteStringConversion(n int) {
	s := strings.Repeat("hello ", 100)
	start := time.Now()
	for i := 0; i < n; i++ {
		b := []byte(s)   // copy 1: string -> []byte
		_ = string(b)    // copy 2: []byte -> string
	}
	fmt.Printf("  %d string<->[]byte round-trips: %s\n", n, time.Since(start))
}

func main() {
	n := 50_000

	fmt.Println("=== Strings, Bytes, and Builders ===")
	fmt.Println()
	fmt.Printf("Building %d-element string...\n\n", n)

	start := time.Now()
	s1 := concatPlus(n)
	t1 := time.Since(start)

	start = time.Now()
	s2 := concatBuilder(n)
	t2 := time.Since(start)

	start = time.Now()
	s3 := concatJoin(n)
	t3 := time.Since(start)

	fmt.Printf("  += concat:        %s  (len=%d)\n", t1, len(s1))
	fmt.Printf("  strings.Builder:  %s  (len=%d)\n", t2, len(s2))
	fmt.Printf("  strings.Join:     %s  (len=%d)\n", t3, len(s3))
	fmt.Println()

	// Show byte<->string conversion cost
	fmt.Println("--- []byte <-> string conversion cost ---")
	byteStringConversion(1_000_000)
	fmt.Println()

	// fmt.Sprintf vs strconv
	fmt.Println("--- fmt.Sprintf vs strconv ---")
	iters := 1_000_000

	start = time.Now()
	for i := 0; i < iters; i++ {
		_ = fmt.Sprintf("%d", i)
	}
	sprintfTime := time.Since(start)

	start = time.Now()
	for i := 0; i < iters; i++ {
		_ = strconv.Itoa(i)
	}
	strconvTime := time.Since(start)

	fmt.Printf("  fmt.Sprintf: %s\n", sprintfTime)
	fmt.Printf("  strconv.Itoa: %s\n", strconvTime)
	fmt.Println()

	fmt.Println("Key: strings.Builder is 10-100x faster than += in loops.")
	fmt.Println("     strconv is 2-5x faster than fmt.Sprintf for numbers.")
}

File: go-backend-interview-prep/09_performance_and_profiling/07_strings_bytes_and_builders/main.py
"""Strings, bytes, and builders -- Python equivalent of the Go example.

Compares string building approaches: +=, list+join, io.StringIO.
"""

import io
import time


def concat_plus(n: int) -> str:
    """Slow in general (CPython may optimize via refcount hack)."""
    s = ""
    for i in range(n):
        s += f"item{i},"
    return s


def concat_join(n: int) -> str:
    """Fast: build list, join once at the end."""
    parts = [f"item{i}" for i in range(n)]
    return ",".join(parts)


def concat_stringio(n: int) -> str:
    """io.StringIO: buffer-based, similar to Go's strings.Builder."""
    buf = io.StringIO()
    for i in range(n):
        buf.write(f"item{i},")
    return buf.getvalue()


def bytes_str_conversion(n: int) -> None:
    """Show cost of encoding/decoding between bytes and str."""
    s = "hello " * 100
    start = time.perf_counter()
    for _ in range(n):
        b = s.encode("utf-8")   # str -> bytes
        _ = b.decode("utf-8")   # bytes -> str
    elapsed = time.perf_counter() - start
    print(f"  {n} str<->bytes round-trips: {elapsed:.4f}s")


def main() -> None:
    n = 50_000

    print("=== Strings, Bytes, and Builders ===\n")
    print(f"Building {n}-element string...\n")

    start = time.perf_counter()
    s1 = concat_plus(n)
    t1 = time.perf_counter() - start

    start = time.perf_counter()
    s2 = concat_join(n)
    t2 = time.perf_counter() - start

    start = time.perf_counter()
    s3 = concat_stringio(n)
    t3 = time.perf_counter() - start

    print(f"  += concat:     {t1:.4f}s  (len={len(s1)})")
    print(f"  ''.join():     {t2:.4f}s  (len={len(s2)})")
    print(f"  io.StringIO:   {t3:.4f}s  (len={len(s3)})")
    print()

    # bytes <-> str conversion cost
    print("--- bytes <-> str conversion cost ---")
    bytes_str_conversion(500_000)
    print()

    # f-string vs % formatting vs str()
    print("--- Formatting comparison ---")
    iters = 500_000

    start = time.perf_counter()
    for i in range(iters):
        _ = f"{i}"
    fstring_time = time.perf_counter() - start

    start = time.perf_counter()
    for i in range(iters):
        _ = str(i)
    str_time = time.perf_counter() - start

    print(f"  f-string:  {fstring_time:.4f}s")
    print(f"  str():     {str_time:.4f}s")
    print()

    print("Key: ''.join(list) is Python's equivalent of Go's strings.Builder.")
    print("     f-strings are fast and readable -- use them by default.")


if __name__ == "__main__":
    main()
